crates/ruff_linter/src/rules/ruff/rules/unreachable.rs

```diff
@@ -0,0 +1,1354 @@
+        case: &'stmt MatchCase,
+    },
+    // Exception was raised and caught by `except` clause
+    ExceptionCaught(&'stmt Expr),
+    // Exception was raised in a `try` block
+    ExceptionRaised,
+}
+
+impl<'stmt> Ranged for Condition<'stmt> {
+    fn range(&self) -> TextRange {
+        match self {
+            Condition::Test(expr)
+            | Condition::Iterator(expr)
+            | Condition::ExceptionCaught(expr) => expr.range(),
+            // The case of the match statement, without the body.
+            Condition::Match { subject: _, case } => TextRange::new(
+                case.start(),
+                case.guard
+                    .as_ref()
+                    .map_or(case.pattern.end(), |guard| guard.end()),
+            ),
+            Condition::ExceptionRaised => TextRange::new(TextSize::new(0), TextSize::new(0)),
+        }
+    }
+}
+
+impl<'stmt> BasicBlock<'stmt> {
+    /// A sentinel block indicating an empty termination block.
+    const EMPTY: BasicBlock<'static> = BasicBlock {
+        stmts: &[],
+        next: NextBlock::Terminate,
+    };
+
+    /// A sentinel block indicating an exception was raised.
+    const EXCEPTION: BasicBlock<'static> = BasicBlock {
+        stmts: &[Stmt::Return(StmtReturn {
+            range: TextRange::new(TextSize::new(0), TextSize::new(0)),
+            value: None,
+        })],
+        next: NextBlock::Terminate,
+    };
+
+    const LOOP_CONTINUE: BasicBlock<'static> = BasicBlock {
+        stmts: &[Stmt::Continue(StmtContinue {
+            range: TextRange::new(TextSize::new(0), TextSize::new(0)),
+        })],
+        next: NextBlock::Terminate, // This must be updated dynamically
+    };
+
+    /// Return true if the block is a sentinel or fake block.
+    fn is_sentinel(&self) -> bool {
+        self.is_empty() || self.is_exception() || self.is_loop_continue()
+    }
+
+    /// Returns true if `self` is an empty block that terminates.
+    fn is_empty(&self) -> bool {
+        matches!(self.next, NextBlock::Terminate) && self.stmts.is_empty()
+    }
+
+    /// Returns true if `self` is a [`BasicBlock::EXCEPTION`].
+    fn is_exception(&self) -> bool {
+        matches!(self.next, NextBlock::Terminate) && BasicBlock::EXCEPTION.stmts == self.stmts
+    }
+
+    /// Returns true if `self` is a [`BasicBlock::LOOP_CONTINUE`].
+    fn is_loop_continue(&self) -> bool {
+        self.stmts == BasicBlock::LOOP_CONTINUE.stmts
+    }
+}
+
+/// Handle a loop block, such as a `while`, `for` or `async for` statement.
+fn loop_block<'stmt>(
+    blocks: &mut BasicBlocksBuilder<'stmt>,
+    condition: Condition<'stmt>,
+    body: &'stmt [Stmt],
+    orelse: &'stmt [Stmt],
+    after: Option<BlockIndex>,
+) -> NextBlock<'stmt> {
+    let after_block = blocks.maybe_next_block_index(after, || true);
+    let last_orelse_statement = blocks.append_blocks_if_not_empty(orelse, after_block);
+
+    let loop_continue_index = blocks.blocks.push(BasicBlock::LOOP_CONTINUE.clone());
```

Comment:
`continue` can and does link to `block 0`, the point of the comment is that this is "odd" since it implies that both `i = 0` and `while True` are executed after each loop, which isn't true but isn't important for the purpose of determining reachability. This behaviour can be changed easily if desired so that `i = 0` and `while True` are seperate blocks.